Add hop annotation and archiving to traceroute-caller

SaiedKazemi opened the pull request:
This commit is essentially based on gfr@'s PR#105 and introduces a
new cache called HopCache. While the implementation isn't fully
complete yet (e.g., the hop cache isn't reset at midnight),
it can be reviewed before getting too large.

Passes all tests via "go test ./..." and also works with
docker-compose.  The annotation files are empty most likely
because there is no database on my local workstation.

	{
	  "Geo": {
	    "Missing": true
	  },
	  "Network": {
	    "Missing": true
	  }
	}

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/m-lab/traceroute-caller/117)
<!-- Reviewable:end -->

Commit:
Add hop annotation and archiving to traceroute-caller

This commit is essentially based on gfr@'s PR#105 and introduces a
new cache called HopCache. While the implementation isn't fully
complete yet (e.g., the hop cache isn't reset at midnight),
it can be reviewed before getting too large.

Passes all tests via "go test ./..." and also works with
docker-compose.  The annotation files are empty most likely
because there is no database on my local workstation.

	{
	  "Geo": {
	    "Missing": true
	  },
	  "Network": {
	    "Missing": true
	  }
	}

stephen-soltesz reviewed (commented):
Cool. Added a few comments.

> *[Reviewable](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-:-Mf9auCR0mjNRJd_UaWO:bwd6dzn)* status: 0 of 1 approvals obtained (waiting on @gfr10598 and @SaiedKazemi)

___
*[hopannotation/hopannotation.go, line 19 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-Mf9Yrr15NRZEsBQCyfk:-Mf9Yrr15NRZEsBQCyfl:blsy27w) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/hopannotation/hopannotation.go#L19)):*
> ```Go
> 	"cloud.google.com/go/civil"
> 
> 	// TODO these should both be in an m-lab/api repository containing only API defs.
> ```

Please rephrase to "a common location containing API defs". Specifying `m-lab/api` assumes an implementation strategy that has not yet been agreed on.
___
*[hopannotation/hopannotation.go, line 94 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-Mf9_8iQ4wr4PBfEMfGK:-Mf9_8iR-hx5yaE_Of5w:b-qsm0xq) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/hopannotation/hopannotation.go#L94)):*
> ```Go
> 		if net.ParseIP(ip) == nil {
> 			log.Printf("AnnorateNewHops(): invalid IP address ip=%v\n", ip)
> 			return 0, 0, fmt.Errorf("invalid IP address")
> ```

No wrong answer -- should the call abort or continue?

Also, it looks like `ExtractHops` already performs a `net.ParseIP` -- is this second check necessary here?
___
*[hopannotation/hopannotation.go, line 114 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-Mf9aVG22MNXTcBEcdQN:-Mf9aVG3AL82DngjsIZc:b-x7u635) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/hopannotation/hopannotation.go#L114)):*
> ```Go
> 	// Aggregate and return errors and counts.
> 	for ip, annotation := range annotations {
> 		filename, err := generateFilename(hc.outputPath, timestamp, uuid, ip)
> ```

Is this the correct filename format for hop annotations?


<!-- Sent from Reviewable.io -->

gfr10598 reviewed (commented):
Adding comments in connectionlistener and caller.go.  I'll add more in other places shortly.

> *[Reviewable](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-:-MfAKFaK0KbfxDlb6UHE:b-rxmkhj)* status: 0 of 1 approvals obtained (waiting on @SaiedKazemi)

___
*[caller.go, line 137 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfAJfdP8Pa50MwJhrDX:-MfAJfdP8Pa50MwJhrDY:bmlyum9) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/caller.go#L137)):*
> ```Go
> 	} else {
> 		go func() {
> 			connCreator, err := connection.NewCreator()
> ```

FYI xxxCreator is a bad name.  It is not creating connections.  It is creating connection objects from SockIDs, IIRC.
___
*[connectionlistener/connectionlistener.go, line 26 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfAIog7BAyk-Qsplyet:-MfAIog7BAyk-Qsplyeu:b-zhjcnt) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/connectionlistener/connectionlistener.go#L26)):*
> ```Go
> // to build a simple client for that service using the libraries from the
> // service itself.
> type connectionListener struct {
> ```

Maybe name this connectionEventHandler?
___
*[connectionlistener/connectionlistener.go, line 43 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfAJwIJ73sJvWTSb-7w:-MfAJwIJ73sJvWTSb-7x:br15m96) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/connectionlistener/connectionlistener.go#L43)):*
> ```Go
> 	defer cl.mutex.Unlock()
> 	if id != nil {
> 		conn, err := cl.creator.FromSockID(*id)
> ```

If we move the action to Open, this FromSockID can be moved into the tracing func that I suggest extracting down below.  The handler really needn't know anything about connections - just pass the uuid and SockID and let the later code sort it out.
___
*[connectionlistener/connectionlistener.go, line 55 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfAIWu_DWc7lUqRtCIf:-MfAIWu_DWc7lUqRtCIg:b-1b1aqj) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/connectionlistener/connectionlistener.go#L55)):*
> ```Go
> // XXX This function should return error instead of ignoring it
> //     and let the caller decide to ignore or handle errors.
> func (cl *connectionListener) Close(ctx context.Context, timestamp time.Time, uuid string) {
> ```

I think now is a fine time to move functionality to Open, but if you prefer we can leave that for next PR.
___
*[connectionlistener/connectionlistener.go, line 67 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfAIe-oDflTKbW7kSrQ:-MfAIe-oDflTKbW7kSrR:b-k8ojkn) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/connectionlistener/connectionlistener.go#L67)):*
> ```Go
> 	}
> 
> 	go func() {
> ```

Suggest we extract this function and name it something meaningful.
___
*[connectionlistener/connectionlistener.go, line 72 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfAJ19H2zbqPYa5wqrR:-MfAJ19H2zbqPYa5wqrS:bth08v2) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/connectionlistener/connectionlistener.go#L72)):*
> ```Go
> 			log.Printf("failed to trace %v", conn.RemoteIP)
> 			return
> 		}
> ```

I suggest the following outline, per our discussion earlier: 
  ...
  hops, err := extractHops(...)
  newHops := findNewHops(hops)
  for i := range newHops {
    err := createAnnotation(newHops[i])
    handle err ...
}
  


<!-- Sent from Reviewable.io -->

gfr10598 reviewed (commented):
Expressing here my concern/ambivalence about adding function into messy structure, instead of refactoring first.  I think we should collapse the connection.* code soon!

> *[Reviewable](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-:-MfAKw2u-ylz10Vo1X_K:b-1h9pnr)* status: 0 of 1 approvals obtained (waiting on @SaiedKazemi)

___
*[caller.go, line 106 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfAKmsf3wnFFhbsHqPs:-MfAKmsgDTcSvJalBQoc:bigv7z) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/caller.go#L106)):*
> ```Go
> 
> 	// Set up the ipCache depending on the trace method requested.
> 	var ipCache *ipcache.RecentIPCache
> ```

Do we want the ipCache exposed this high up in the call stack?  I'm not sure of the answer.
___
*[caller.go, line 140 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfALERW3Yxwyt5bGAIG:-MfALERW3Yxwyt5bGAIH:b-h4p3rl) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/caller.go#L140)):*
> ```Go
> 			rtx.Must(err, "failed to discover local IPs")
> 			ipserviceClient := ipservice.NewClient(*ipservice.SocketFilename)
> 			hopCache := hopannotation.New(ipserviceClient, *outputPath)
> ```

hopAnnotator ?  The cache is an implementation detail, so it seems odd to name it that.  (Same for ipcache, which we can address later).
___
*[connectionlistener/connectionlistener.go, line 67 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfAIe-oDflTKbW7kSrQ:-MfARU2rEdrM8PJUjD6G:b-bifh40) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/connectionlistener/connectionlistener.go#L67)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Suggest we extract this function and name it something meaningful.
</blockquote></details>

Maybe TraceAndAnnotate, or TraceAnnotateArchive?
___
*[connectionlistener/connectionlistener.go, line 72 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfAJ19H2zbqPYa5wqrR:-MfARPKL9Gsvsdfy-rHF:b-b2aq36) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/connectionlistener/connectionlistener.go#L72)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

I suggest the following outline, per our discussion earlier: 
  ...
  hops, err := extractHops(...)
  newHops := findNewHops(hops)
  for i := range newHops {
    err := createAnnotation(newHops[i])
    handle err ...
}
  
</blockquote></details>

Grr - whitespace ignored.
___
*[hopannotation/hopannotation.go, line 148 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfALqMi04L4TnITT_Ai:-MfALqMi04L4TnITT_Aj:b-tfrwb8) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/hopannotation/hopannotation.go#L148)):*
> ```Go
> 
> // archiveHop archives the given hop IP. The archive filename should be in the
> // "<date>_<machine-site>_<ip>.json" format.
> ```

This comment needs to be updated.
___
*[hopannotation/hopannotation.go, line 149 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfALwjGB3RnonY6MmW6:-MfALwjGB3RnonY6MmW7:b-98hi5x) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/hopannotation/hopannotation.go#L149)):*
> ```Go
> // archiveHop archives the given hop IP. The archive filename should be in the
> // "<date>_<machine-site>_<ip>.json" format.
> func archiveHop(ctx context.Context, filename string, annotation *annotator.ClientAnnotations) error {
> ```

We should move toward either separating the generation and the writing, or, inject a Writer into the generator.  Injecting a filename is very limiting.

There is a m-lab/go/content.Provider, which has been useful for content sources.  We could create something similar, but I think WriteCloser interface should be fine.  We could implement a gcsWriteCloser that holds a path and other attributes, does nothing until Write is called, and cleans everything up when Close is called.  Or we could write Consumer, that just takes a single WriteAll([]byte) and automatically does the create and close and returns any errors.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[caller.go, line 137 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfAJfdP8Pa50MwJhrDX:-MfASK4Z-BcALXlqInTn:b-nkif2s) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/caller.go#L137)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

FYI xxxCreator is a bad name.  It is not creating connections.  It is creating connection objects from SockIDs, IIRC.
</blockquote></details>

Yeah, it's a bad name but I chose it because we get this object from `NewCreator()`.  We need to change `NewCreator()` to something better and also have it return a concrete type as opposed to an interface.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[caller.go, line 137 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfAJfdP8Pa50MwJhrDX:-MfATGVLAL3y8fOysWzs:bw23jfd) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/caller.go#L137)):*
<details><summary><i>Previously, SaiedKazemi (Saied Kazemi) wrote…</i></summary><blockquote>

Yeah, it's a bad name but I chose it because we get this object from `NewCreator()`.  We need to change `NewCreator()` to something better and also have it return a concrete type as opposed to an interface.
</blockquote></details>

By the way, `NewCreator()` does not create connection objects from SockIDs.  Its comment says `NewCreator makes an object that can convert src and dst into local and remote IPs.`


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[caller.go, line 106 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfAKmsf3wnFFhbsHqPs:-MfAUD3u7OJXloDw4ptG:bgm1zsk) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/caller.go#L106)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Do we want the ipCache exposed this high up in the call stack?  I'm not sure of the answer.
</blockquote></details>

This is the original code.  I just changed the name from `cache` to `ipCache` for more clarity.

Re whether it should be visible at this level or not, let's consider the question when we refactor TRC.  There's also a lot of other questions like this that we need to answer, so (as we discussed) let's limit the scope of current work to adding hop annotations.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[caller.go, line 140 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfALERW3Yxwyt5bGAIG:-MfBdCRtAmjRL7Etwvcl:b-896fix) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/caller.go#L140)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

hopAnnotator ?  The cache is an implementation detail, so it seems odd to name it that.  (Same for ipcache, which we can address later).
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[connectionlistener/connectionlistener.go, line 26 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfAIog7BAyk-Qsplyet:-MfBdJsyERP6JiRD0CMA:bxqz6te) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/connectionlistener/connectionlistener.go#L26)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Maybe name this connectionEventHandler?
</blockquote></details>

This is the original code.  Let's change the names when we do the refactoring.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[connectionlistener/connectionlistener.go, line 43 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfAJwIJ73sJvWTSb-7w:-MfBdl01Cqn3l2_5dFsc:b-g8g3yl) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/connectionlistener/connectionlistener.go#L43)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

If we move the action to Open, this FromSockID can be moved into the tracing func that I suggest extracting down below.  The handler really needn't know anything about connections - just pass the uuid and SockID and let the later code sort it out.
</blockquote></details>

Sounds good, but let's do that when we refactor the code so the scope of this PR remains adding hop annotation functionality.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[connectionlistener/connectionlistener.go, line 55 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfAIWu_DWc7lUqRtCIf:-MfBe47C6EzdoEdEk8ov:b-92oe9) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/connectionlistener/connectionlistener.go#L55)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

I think now is a fine time to move functionality to Open, but if you prefer we can leave that for next PR.
</blockquote></details>

Yeah, let's leave it for the refactoring PR.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[connectionlistener/connectionlistener.go, line 67 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfAIe-oDflTKbW7kSrQ:-MfBgPCZD8Qiv-F6KXAu:b-896fix) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/connectionlistener/connectionlistener.go#L67)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Maybe TraceAndAnnotate, or TraceAnnotateArchive?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[connectionlistener/connectionlistener.go, line 72 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfAJ19H2zbqPYa5wqrR:-MfBiogzAa-z-8Z6-kgZ:b-896fix) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/connectionlistener/connectionlistener.go#L72)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Grr - whitespace ignored.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 19 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-Mf9Yrr15NRZEsBQCyfk:-MfBjKsv6_gEwqgTPtdE:b-896fix) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/hopannotation/hopannotation.go#L19)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Please rephrase to "a common location containing API defs". Specifying `m-lab/api` assumes an implementation strategy that has not yet been agreed on.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 94 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-Mf9_8iQ4wr4PBfEMfGK:-MfBm2ZZEb0X5tLDH9NZ:bs2w0gm) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/hopannotation/hopannotation.go#L94)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

No wrong answer -- should the call abort or continue?

Also, it looks like `ExtractHops` already performs a `net.ParseIP` -- is this second check necessary here?
</blockquote></details>

I fixed the check in `ExtractHops()` and removed the check here.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 114 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-Mf9aVG22MNXTcBEcdQN:-MfBowJxCS6_O635vUIU:b-9en6ro) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/hopannotation/hopannotation.go#L114)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Is this the correct filename format for hop annotations?
</blockquote></details>

Per our conversation, I have changed it to "<yyyymmdd>_<hostname>_<ip>.json".


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 149 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfALwjGB3RnonY6MmW6:-MfBrc_G85XJ0Qnp5sHS:bjzpxnc) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/hopannotation/hopannotation.go#L149)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

We should move toward either separating the generation and the writing, or, inject a Writer into the generator.  Injecting a filename is very limiting.

There is a m-lab/go/content.Provider, which has been useful for content sources.  We could create something similar, but I think WriteCloser interface should be fine.  We could implement a gcsWriteCloser that holds a path and other attributes, does nothing until Write is called, and cleans everything up when Close is called.  Or we could write Consumer, that just takes a single WriteAll([]byte) and automatically does the create and close and returns any errors.
</blockquote></details>

Acknowledged.  Let's discuss interactively and make the change in a subsequent PR.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 148 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfALqMi04L4TnITT_Ai:-MfBrMZJ5QKokWktqPJG:b-896fix) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/hopannotation/hopannotation.go#L148)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

This comment needs to be updated.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

Commit:
Make changes from code reviews

After installing real annotation databases on my workstation,
I am now getting valid hop annotations. For example:

    $ jq . < 20210722_565e36f9fe3d_91.189.88.152.json
    {
      "Geo": {
        "ContinentCode": "EU",
        "CountryCode": "GB",
        "CountryName": "United Kingdom",
        "Subdivision1ISOCode": "ENG",
        "Subdivision1Name": "England",
        "City": "London",
        "PostalCode": "EC2V",
        "Latitude": 51.5095,
        "Longitude": -0.0955,
        "AccuracyRadiusKm": 200
      },
      "Network": {
        "CIDR": "91.189.88.0/21",
        "ASNumber": 41231,
        "ASName": "Canonical Group Limited",
        "Systems": [
          {
            "ASNs": [
              41231
            ]
          }
        ]
      }
    }

stephen-soltesz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/m-lab/traceroute-caller/117)* status: 0 of 1 approvals obtained (waiting on @gfr10598 and @SaiedKazemi)

___
*[hopannotation/hopannotation.go, line 94 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-Mf9_8iQ4wr4PBfEMfGK:-MfDdUSYAq83x3Z_YpZ4:b-pgox8t) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/hopannotation/hopannotation.go#L94)):*
<details><summary><i>Previously, SaiedKazemi (Saied Kazemi) wrote…</i></summary><blockquote>

I fixed the check in `ExtractHops()` and removed the check here.
</blockquote></details>

I'm confused. Do both functions need to parse the hop IPs? Is this still for debugging?
___
*[hopannotation/hopannotation.go, line 114 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-Mf9aVG22MNXTcBEcdQN:-MfDdrfWAUhwkycx3pJq:b-jr86ke) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/hopannotation/hopannotation.go#L114)):*
<details><summary><i>Previously, SaiedKazemi (Saied Kazemi) wrote…</i></summary><blockquote>

Per our conversation, I have changed it to "<yyyymmdd>_<hostname>_<ip>.json".
</blockquote></details>

I think the formatter ate your formatting! `"<yyyymmdd>_<hostname>_<ip>.json".`


<!-- Sent from Reviewable.io -->

stephen-soltesz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/m-lab/traceroute-caller/117)* status: 0 of 1 approvals obtained (waiting on @gfr10598 and @SaiedKazemi)

___
*[hopannotation/hopannotation.go, line 149 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfALwjGB3RnonY6MmW6:-MfDgz0qEiVjKZS_tF5A:bih8vkj) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/hopannotation/hopannotation.go#L149)):*
<details><summary><i>Previously, SaiedKazemi (Saied Kazemi) wrote…</i></summary><blockquote>

Acknowledged.  Let's discuss interactively and make the change in a subsequent PR.
</blockquote></details>

@gfr10598 I don't understand your comment -- what is limiting about filenames in this context? When will traceroute-caller write to a non-local location?


<!-- Sent from Reviewable.io -->

gfr10598 reviewed (commented):
> Reviewed 1 of 10 files at r1, 4 of 8 files at r2.
> *[Reviewable](https://reviewable.io/reviews/m-lab/traceroute-caller/117)* status: 0 of 1 approvals obtained (waiting on @gfr10598, @SaiedKazemi, and @stephen-soltesz)

___
*[caller.go, line 137 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfAJfdP8Pa50MwJhrDX:-MfDZTVWBsNGXM0zvN7S:b-7b819p) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/caller.go#L137)):*
<details><summary><i>Previously, SaiedKazemi (Saied Kazemi) wrote…</i></summary><blockquote>

By the way, `NewCreator()` does not create connection objects from SockIDs.  Its comment says `NewCreator makes an object that can convert src and dst into local and remote IPs.`
</blockquote></details>

Yup - I meant connCreator, not NewCreator.  thanks for fixing!
___
*[connectionlistener/connectionlistener.go, line 1 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfDcbQi3lrqCoDE2tC0:-MfDcbQi3lrqCoDE2tC1:b-gqn82n) ([raw file](https://github.com/m-lab/traceroute-caller/blob/379fd9fdf0e59b94a1bdaaae3deef701d5d90970/connectionlistener/connectionlistener.go#L1)):*
> ```Go
> // Package connectionlistener provides two handles for open and close
> ```

s/handles/handlers/
___
*[connectionlistener/connectionlistener.go, line 66 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfDcz2X-0TQmScDLfBJ:-MfDcz2X-0TQmScDLfBK:b-ganz37) ([raw file](https://github.com/m-lab/traceroute-caller/blob/379fd9fdf0e59b94a1bdaaae3deef701d5d90970/connectionlistener/connectionlistener.go#L66)):*
> ```Go
> 
> // traceAnnotateArchive runs a traceroute, takes its output, extracts all tracelb's hop
> // IP addresses, annotates the IPs, and archives the annotations (writes to a file).
> ```

writes to one or more files.
___
*[connectionlistener/connectionlistener.go, line 96 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfDdCva5j2w9HCTNuLT:-MfDdCva5j2w9HCTNuLU:b-upi7la) ([raw file](https://github.com/m-lab/traceroute-caller/blob/379fd9fdf0e59b94a1bdaaae3deef701d5d90970/connectionlistener/connectionlistener.go#L96)):*
> ```Go
> 		return
> 	}
> 	if len(hops) == 0 {
> ```

We should add a histogram metric for hops, prior to deployment to staging I think.  We can do that in parallel with the refactoring work.
___
*[connectionlistener/connectionlistener_test.go, line 77 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfDe1Y25H7P8EALwrqD:-MfDe1Y25H7P8EALwrqE:b-gghofa) ([raw file](https://github.com/m-lab/traceroute-caller/blob/379fd9fdf0e59b94a1bdaaae3deef701d5d90970/connectionlistener/connectionlistener_test.go#L77)):*
> ```Go
> 
> 	localIP := net.ParseIP("10.0.0.1")
> 	creator := connection.NewFakeCreator([]*net.IP{&localIP})
> ```

This name will come back to bite us.  Rename NewFakeLocalIP?
___
*[hopannotation/hopannotation.go, line 114 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-Mf9aVG22MNXTcBEcdQN:-MfDfJtI393stLY0AVjH:b-ttbh3w) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/hopannotation/hopannotation.go#L114)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

I think the formatter ate your formatting! `"<yyyymmdd>_<hostname>_<ip>.json".`
</blockquote></details>

I guess you discussed including the TxxxxxxZ?  I think that even though we don't need it, it would be nice to maintain this part of the pattern used in all the other filenames.  

We don't need the hostname.  That is included in the archive name, which is how we handle it for all the other datatypes, IIRC.  Discussed with Saied yesterday, but not sure if you talked to Stephen since then.

So, I propose <yyyymmddThhmmss.xxxxxxZ_ip.json

We also include a <pod name>_<boottime> in other names, ndt-ps5k9_1620781375 which seems quite useful as well.  ndt7 also include ndt7-upload, which is probably not necessary, as the pusher should give us that info.

So, maybe
 hop-annotation_<pod name>_<boottime>_<yyyymmddThhmmss.xxxxxxZ>_<ip>.json

I think we could reasonably get by with either <podname>_<boottime> OR the timestamp, but without one of those, it will be harder to disambiguate annotations before and after a reboot.  I guess we'd have to use the file creation time, which I hate to rely on.
___
*[hopannotation/hopannotation.go, line 27 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfDeI-GC0u9LALGJiuk:-MfDeI-GC0u9LALGJiul:b-c4e3wy) ([raw file](https://github.com/m-lab/traceroute-caller/blob/379fd9fdf0e59b94a1bdaaae3deef701d5d90970/hopannotation/hopannotation.go#L27)):*
> ```Go
> var (
> 	// HopArchiver is for blackbox package testing.
> 	HopArchiver = archiveHopAnnotation
> ```

This should be exposed in an export_test.go file,


<!-- Sent from Reviewable.io -->

gfr10598 reviewed (commented):
> Reviewed 1 of 10 files at r1, 2 of 8 files at r2.
> *[Reviewable](https://reviewable.io/reviews/m-lab/traceroute-caller/117)* status: 0 of 1 approvals obtained (waiting on @gfr10598, @SaiedKazemi, and @stephen-soltesz)



<!-- Sent from Reviewable.io -->

gfr10598 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/m-lab/traceroute-caller/117)* status: 0 of 1 approvals obtained (waiting on @gfr10598, @SaiedKazemi, and @stephen-soltesz)

___
*[hopannotation/hopannotation.go, line 140 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfJGyB_4nfefSVEyZQV:-MfJGyB_4nfefSVEyZQW:b-gbxwma) ([raw file](https://github.com/m-lab/traceroute-caller/blob/379fd9fdf0e59b94a1bdaaae3deef701d5d90970/hopannotation/hopannotation.go#L140)):*
> ```Go
> 	dir := dirPath + "/" + timestamp.Format("2006/01/02")
> 	if err := os.MkdirAll(dir, 0777); err != nil {
> 		return "", errors.New("could not create output directory") // TODO add metric here
> ```

Create an exported ErrXXX var and use it here.  That makes it possible to test against in other code, and helps document all the errors that might be encountered.


<!-- Sent from Reviewable.io -->

gfr10598 reviewed (commented):
FYI I ran this manually in sandbox Friday afternoon, using this fragment:
       - -ipservice.sock=/var/local/uuidannotatorsocket/annotator.sock
        env:
        - name: PRIVATE_IP
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: status.podIP
        image: measurementlab/traceroute-caller:sandbox-saied-hopannotation

It is generating errors for some hop ips, like this:
2021/07/23 17:08:32 connectionlistener.go:104: failed to annotate and archive 34.135.1.66 (n=0 k=0 err=Got HTTP 400, but wanted HTTP 200)
2021/07/23 17:08:32 hopannotation.go:91: failed to annotate hops (error: Got HTTP 400, but wanted HTTP 200)
2021/07/23 17:08:32 connectionlistener.go:104: failed to annotate and archive 34.135.1.66 (n=0 k=0 err=Got HTTP 400, but wanted HTTP 200)
2021/07/23 17:08:44 hopannotation.go:91: failed to annotate hops (error: Got HTTP 400, but wanted HTTP 200)
2021/07/23 17:08:44 connectionlistener.go:104: failed to annotate and archive 34.135.1.66 (n=0 k=0 err=Got HTTP 400, but wanted HTTP 200)
2021/07/23 17:08:49 hopannotation.go:91: failed to annotate hops (error: Got HTTP 400, but wanted HTTP 200)
2021/07/23 17:08:49 connectionlistener.go:104: failed to annotate and archive 34.135.1.66 (n=0 k=0 err=Got HTTP 400, but wanted HTTP 200)
2021/07/23 17:09:32 scamper.go:259: Trace returned in 270.364147623 seconds
2021/07/23 17:09:32 scamper.go:275: Trace succeeded in context 0xc00042e780
2021/07/23 17:09:32 hopannotation.go:91: failed to annotate hops (error: Got HTTP 400, but wanted HTTP 200)
2021/07/23 17:09:32 connectionlistener.go:104: failed to annotate and archive 35.184.60.232 (n=0 k=0 err=Got HTTP 400, but wanted HTTP 200)
2021/07/23 17:09:32 hopannotation.go:91: failed to annotate hops (error: Got HTTP 400, but wanted HTTP 200)
2021/07/23 17:09:32 connectionlistener.go:104: failed to annotate and archive 35.184.60.232 (n=0 k=0 err=Got HTTP 400, but wanted HTTP 200)
2021/07/23 17:09:32 hopannotation.go:91: failed to annotate hops (error: Got HTTP 400, but wanted HTTP 200)
2021/07/23 17:09:32 connectionlistener.go:104: failed to annotate and archive 35.184.60.232 (n=0 k=0 err=Got HTTP 400, but wanted HTTP 200)
2021/07/23 17:09:32 hopannotation.go:91: failed to annotate hops (error: Got HTTP 400, but wanted HTTP 200)
2021/07/23 17:09:32 connectionlistener.go:104: failed to annotate and archive 35.184.60.232 (n=0 k=0 err=Got HTTP 400, but wanted HTTP 200)
2021/07/23 17:09:32 hopannotation.go:91: failed to annotate hops (error: Got HTTP 400, but wanted HTTP 200)
2021/07/23 17:09:32 connectionlistener.go:104: failed to annotate and archive 35.184.60.232 (n=0 k=0 err=Got HTTP 400, but wanted HTTP 200)
2021/07/23 17:09:44 hopannotation.go:91: failed to annotate hops (error: Got HTTP 400, but wanted HTTP 200)
2021/07/23 17:09:44 connectionlistener.go:104: failed to annotate and archive 35.184.60.232 (n=0 k=0 err=Got HTTP 400, but wanted HTTP 200)


 

> *[Reviewable](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-:-MfJJbov45FzUi2qZBWF:b-r7ogeo)* status: 0 of 1 approvals obtained (waiting on @SaiedKazemi and @stephen-soltesz)

___
*[hopannotation/hopannotation.go, line 91 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfJIv2O0D7nDmRSy7KX:-MfJIv2P7cBIYNS2bfSs:bg3si7m) ([raw file](https://github.com/m-lab/traceroute-caller/blob/379fd9fdf0e59b94a1bdaaae3deef701d5d90970/hopannotation/hopannotation.go#L91)):*
> ```Go
> 	annotations, err := hc.annotator.Annotate(context.Background(), newHops) // uuid-annotator/ipservice/client.go:Annotate()
> 	if err != nil {
> 		log.Printf("failed to annotate hops (error: %v)\n", err)
> ```

This is producing redundant log messages, alongsize connectionlistener.go:104


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 94 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-Mf9_8iQ4wr4PBfEMfGK:-MfYxQuc9AyXaOJBvvfH:b-cazkhm) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/hopannotation/hopannotation.go#L94)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

I'm confused. Do both functions need to parse the hop IPs? Is this still for debugging?
</blockquote></details>

I removed this but put it back in because package testing failed due to an invalid IP address that was intentionally passed in.  More importantly, because this is an exported method and can be called from other callers, it needs to validate its input.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 91 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfJIv2O0D7nDmRSy7KX:-MfYzHdIBG1DF61LTc-8:b-jc4k45) ([raw file](https://github.com/m-lab/traceroute-caller/blob/379fd9fdf0e59b94a1bdaaae3deef701d5d90970/hopannotation/hopannotation.go#L91)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

This is producing redundant log messages, alongsize connectionlistener.go:104
</blockquote></details>

Error handling needs major refactoring in traceroute-caller.  Lower layers should return (not log) errors to higher layers.  Higher layers should judiciously decide whether to ignore the error or wrap the error with more context and pass it up the call chain.

As discussed offline, I will refactor error handling 


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 114 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-Mf9aVG22MNXTcBEcdQN:-MfZ08-H91t1OgOEvzLk:b-2zk11x) ([raw file](https://github.com/m-lab/traceroute-caller/blob/97692d2162f9c3d369cd08439dd41d141577a7f7/hopannotation/hopannotation.go#L114)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

I guess you discussed including the TxxxxxxZ?  I think that even though we don't need it, it would be nice to maintain this part of the pattern used in all the other filenames.  

We don't need the hostname.  That is included in the archive name, which is how we handle it for all the other datatypes, IIRC.  Discussed with Saied yesterday, but not sure if you talked to Stephen since then.

So, I propose <yyyymmddThhmmss.xxxxxxZ_ip.json

We also include a <pod name>_<boottime> in other names, ndt-ps5k9_1620781375 which seems quite useful as well.  ndt7 also include ndt7-upload, which is probably not necessary, as the pusher should give us that info.

So, maybe
 hop-annotation_<pod name>_<boottime>_<yyyymmddThhmmss.xxxxxxZ>_<ip>.json

I think we could reasonably get by with either <podname>_<boottime> OR the timestamp, but without one of those, it will be harder to disambiguate annotations before and after a reboot.  I guess we'd have to use the file creation time, which I hate to rely on.


</blockquote></details>

Let's finalize this today.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 140 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfJGyB_4nfefSVEyZQV:-MfZ9lxt6He3q6IPf9PY:b5jc8td) ([raw file](https://github.com/m-lab/traceroute-caller/blob/379fd9fdf0e59b94a1bdaaae3deef701d5d90970/hopannotation/hopannotation.go#L140)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Create an exported ErrXXX var and use it here.  That makes it possible to test against in other code, and helps document all the errors that might be encountered.
</blockquote></details>

There is no need for a new error message.  The code should return "err".


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
Will look into this.





<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[connectionlistener/connectionlistener.go, line 66 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfDcz2X-0TQmScDLfBJ:-MfZBHhT8-SPf87kejXw:b-896fix) ([raw file](https://github.com/m-lab/traceroute-caller/blob/379fd9fdf0e59b94a1bdaaae3deef701d5d90970/connectionlistener/connectionlistener.go#L66)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

writes to one or more files.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[connectionlistener/connectionlistener_test.go, line 77 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfDe1Y25H7P8EALwrqD:-MfZCtoq2aG5L8PislCM:b-896fix) ([raw file](https://github.com/m-lab/traceroute-caller/blob/379fd9fdf0e59b94a1bdaaae3deef701d5d90970/connectionlistener/connectionlistener_test.go#L77)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

This name will come back to bite us.  Rename NewFakeLocalIP?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 27 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfDeI-GC0u9LALGJiuk:-MfZD7VCAL7o-4_UIjBg:bvjv19m) ([raw file](https://github.com/m-lab/traceroute-caller/blob/379fd9fdf0e59b94a1bdaaae3deef701d5d90970/hopannotation/hopannotation.go#L27)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

This should be exposed in an export_test.go file,
</blockquote></details>

I don't follow because we don't have `export.go`.  This variable is used in `hopannotation/hopannotation_test.go` so that `AnnotateArchive()` would call `fakeArchiveHopAnnotation()` instead of `archiveHopAnnotation()`.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[connectionlistener/connectionlistener.go, line 1 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfDcbQi3lrqCoDE2tC0:-MfZB0Nr1WB13p3_AOoT:b-896fix) ([raw file](https://github.com/m-lab/traceroute-caller/blob/379fd9fdf0e59b94a1bdaaae3deef701d5d90970/connectionlistener/connectionlistener.go#L1)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

s/handles/handlers/
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

Commit:
Additional changes based on code review.

Commit:
Generate correct hop annotation file name and hop ID

This commit generates hop annotation file name and hop ID as
specified in the design doc:

ID:       <yyyymmdd>_<hostname>_<boottime>_<ip>
Filename: <timestamp>_<hostname>_<boottime>_<ip>.json

Tested locally:

$ pwd
.../local/traceroute/2021/07/28
$ jq . < 20210728T152816Z_6bf574a99207_1627310329_212.187.137.18.json
{
  "ID": "20210728_6bf574a99207_1627310329_212.187.137.18",
  "Date": "2021-07-28",
  "Raw": {
    "Geo": {
      "ContinentCode": "EU",
      "CountryCode": "GB",
      "CountryName": "United Kingdom",
      "Latitude": 51.4964,
      "Longitude": -0.1224,
      "AccuracyRadiusKm": 200
    },
    "Network": {
      "CIDR": "212.187.128.0/17",
      "ASNumber": 3356,
      "ASName": "Level 3 Parent, LLC",
      "Systems": [
        {
          "ASNs": [
            3356
          ]
        }
      ]
    }
  }
}

stephen-soltesz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/m-lab/traceroute-caller/117)* status: 0 of 1 approvals obtained (waiting on @gfr10598 and @stephen-soltesz)

___
*[hopannotation/hopannotation.go, line 27 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfDeI-GC0u9LALGJiuk:-MfiTbt_56ZwvNparZ-n:bjvzz80) ([raw file](https://github.com/m-lab/traceroute-caller/blob/379fd9fdf0e59b94a1bdaaae3deef701d5d90970/hopannotation/hopannotation.go#L27)):*
<details><summary><i>Previously, SaiedKazemi (Saied Kazemi) wrote…</i></summary><blockquote>

I don't follow because we don't have `export.go`.  This variable is used in `hopannotation/hopannotation_test.go` so that `AnnotateArchive()` would call `fakeArchiveHopAnnotation()` instead of `archiveHopAnnotation()`.
</blockquote></details>

Go allows two modes: whitebox testing, blackbox testing. whitebox tests use a package name that matches the primary package under test, in this case, `package hopannotation`. blackbox testing would use a similar package name with a suffix `_test`, so `package hopannotation_test`. This mode only makes the exported methods accessible.

Go also allows (by convention or specification, I'm not sure) exporting private methods only for testing using a file like `export_test.go`.

So, `export_test.go` would import the whitebox package to access the private methods and define new public names for the blackbox tests.

If you search for "export_test.go" you can find examples of this in Go system packages. There are many examples within the m-lab org also.

https://www.google.com/search?q=golang+export_test.go

https://github.com/search?q=org%3Am-lab+filename%3A%22export_test.go%22&type=code

I cannot find upstream documentation for this behavior.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 27 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MfDeI-GC0u9LALGJiuk:-Mfikt801YqGkBK0gEew:bseiwb7) ([raw file](https://github.com/m-lab/traceroute-caller/blob/379fd9fdf0e59b94a1bdaaae3deef701d5d90970/hopannotation/hopannotation.go#L27)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Go allows two modes: whitebox testing, blackbox testing. whitebox tests use a package name that matches the primary package under test, in this case, `package hopannotation`. blackbox testing would use a similar package name with a suffix `_test`, so `package hopannotation_test`. This mode only makes the exported methods accessible.

Go also allows (by convention or specification, I'm not sure) exporting private methods only for testing using a file like `export_test.go`.

So, `export_test.go` would import the whitebox package to access the private methods and define new public names for the blackbox tests.

If you search for "export_test.go" you can find examples of this in Go system packages. There are many examples within the m-lab org also.

https://www.google.com/search?q=golang+export_test.go

https://github.com/search?q=org%3Am-lab+filename%3A%22export_test.go%22&type=code

I cannot find upstream documentation for this behavior.
</blockquote></details>

I am familiar with `whitebox` and `blackbox` testing in Go (`package <pkg>` vs `package <pkg>_test`) but wasn't familiar with the `export_test.go` convention.  IIUC, `export_test.go` is used for exporting private functions of a package for testing (i.e., making them accessible to test code). In our case here, we want to replace a private function with a fake version of it. That said, I will create `export_test.go`.


<!-- Sent from Reviewable.io -->

Commit:
Add debugging messages.

Commit:
Fix the number of expected calls to Annotate()

Commit:
Clean up debug code and hop state to hop cache

Commit:
Add trace output to test data for better coverage

Commit:
Refactor connectionlistener test code for better coverage

Commit:
More changes to get 100% test coverage

Commit:
Refactor package test code for coverage and Go conventions

stephen-soltesz reviewed (commented):
Hey, @saiedkazemi I've added a handful of comments. My main concern at the moment is that this is a very large PR. Are you planning for a time when these changes can be carved into smaller PRs?

> Reviewed 2 of 10 files at r1.
> *[Reviewable](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-:-MggkBZiC00p6C-dMH_A:b-oqqcys)* status: 0 of 1 approvals obtained (waiting on @gfr10598, @SaiedKazemi, and @stephen-soltesz)

___
*[connectionlistener/connectionlistener.go, line 4 at r7](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MggT5Zj0dpyK9TjvTGa:-MggT5Zj0dpyK9TjvTGb:b-7wight) ([raw file](https://github.com/m-lab/traceroute-caller/blob/c904b992180d731601840d82ae1ca84af3e26397/connectionlistener/connectionlistener.go#L4)):*
> ```Go
> // events called for each network connection.
> //
> // XXX The name connectionListener is terribly misleading because this
> ```

Yeah, this is more of an `eventsocket.Handler` implementation - https://github.com/m-lab/tcp-info/blob/master/eventsocket/client.go#L27

Maybe `connectionevent.Handler`?
___
*[hopannotation/hopannotation.go, line 53 at r7](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MggfyaC6MhuGxH-KE5u:-MggfyaC6MhuGxH-KE5v:b-47jip0) ([raw file](https://github.com/m-lab/traceroute-caller/blob/c904b992180d731601840d82ae1ca84af3e26397/hopannotation/hopannotation.go#L53)):*
> ```Go
> 
> // HopAnnotation1 defines the schema for BigQuery hop annotations.
> type HopAnnotation1 struct {
> ```

Please see new suggestions in design doc: https://docs.google.com/document/d/1Kh-YbJnZhm1KhcPFo-qN48wdpxYpNsrKzNLvKF0_0UI/edit?pli=1#heading=h.1o8g4cz8a12n
___
*[hopannotation/hopannotation.go, line 199 at r7](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-Mgg_ftV2ZQpF_PUkhOz:-Mgg_ftV2ZQpF_PUkhP-:b-bp8bvv) ([raw file](https://github.com/m-lab/traceroute-caller/blob/c904b992180d731601840d82ae1ca84af3e26397/hopannotation/hopannotation.go#L199)):*
> ```Go
> 	if err := os.MkdirAll(dirPath, 0777); err != nil {
> 		// TODO: Add a metric here.
> 		return "", fmt.Errorf("%s (error: %v)", errCreatePath, err)
> ```

When combining errors, I believe it's preferable to use `%w` so that callers (within this package) could use `errors.Is` to tell if this is a specific error derived from `errCreatePath`.

EDIT: I see that `errCreatePath` is a string... The naming made me expect that it was an error type.
___
*[parser/json.go, line 19 at r7](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-Mggff9t-IeMQlXlwFY-:-Mggff9t-IeMQlXlwFY0:b3ly8k1) ([raw file](https://github.com/m-lab/traceroute-caller/blob/c904b992180d731601840d82ae1ca84af3e26397/parser/json.go#L19)):*
> ```Go
> 
> var (
> 	errNumLines       = "test has wrong number of lines"
> ```

I did a double take here and in hopannotation.go b/c I thought these variables would be `error` types when reading their usage. Why not make them errors?
___
*[parser/json.go, line 170 at r7](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MggeAKh3mxqxbNiipYp:-MggeAKh3mxqxbNiipYq:byquhzy) ([raw file](https://github.com/m-lab/traceroute-caller/blob/c904b992180d731601840d82ae1ca84af3e26397/parser/json.go#L170)):*
> ```Go
> 
> 	lines := 4
> 	if !strings.HasPrefix(string(data), "{\"UUID\":") {
> ```

Some of this parser code may have been copied from the etl parser and contain more general logic than required for the output created by traceroute-caller directly. Does the current implementation of traceroute-caller create JSONL output with a fixed number of objects? e.g. 4? Is 3 an error?

If this is necessary, is the UUID field guaranteed to be first?
___
*[parser/json.go, line 180 at r7](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-Mggf9064SulPy38h3v2:-Mggf907-sYwfqjo14BW:ba11qm5) ([raw file](https://github.com/m-lab/traceroute-caller/blob/c904b992180d731601840d82ae1ca84af3e26397/parser/json.go#L180)):*
> ```Go
> 
> 	// Validate the cycle-start line.
> 	err := json.Unmarshal([]byte(jsonStrings[n-3]), &cycleStart)
> ```

Again, if the current tool guarantees and output format, then this indexing could be simplified hopefully


<!-- Sent from Reviewable.io -->

Commit:
Changes based on code review and more clean ups

SaiedKazemi commented:
___
*[connectionlistener/connectionlistener.go, line 4 at r7](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MggT5Zj0dpyK9TjvTGa:-MgsxEUrE_qvnYJ4TG9I:b-3dg0oz) ([raw file](https://github.com/m-lab/traceroute-caller/blob/c904b992180d731601840d82ae1ca84af3e26397/connectionlistener/connectionlistener.go#L4)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Yeah, this is more of an `eventsocket.Handler` implementation - https://github.com/m-lab/tcp-info/blob/master/eventsocket/client.go#L27

Maybe `connectionevent.Handler`?
</blockquote></details>

As discussed offline, the three connection packages need refactoring and renaming.  Will do this in a separate PR.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 53 at r7](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MggfyaC6MhuGxH-KE5u:-Mgsx_RE3yUEvp6S5g2U:b-h87i6w) ([raw file](https://github.com/m-lab/traceroute-caller/blob/c904b992180d731601840d82ae1ca84af3e26397/hopannotation/hopannotation.go#L53)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Please see new suggestions in design doc: https://docs.google.com/document/d/1Kh-YbJnZhm1KhcPFo-qN48wdpxYpNsrKzNLvKF0_0UI/edit?pli=1#heading=h.1o8g4cz8a12n
</blockquote></details>

I have made the changes and will share a doc discussing what timestamp we should use for hopannotation1 datatype.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[parser/json.go, line 19 at r7](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-Mggff9t-IeMQlXlwFY-:-MgsxuQqCDQrxa1kngSD:b-896fix) ([raw file](https://github.com/m-lab/traceroute-caller/blob/c904b992180d731601840d82ae1ca84af3e26397/parser/json.go#L19)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

I did a double take here and in hopannotation.go b/c I thought these variables would be `error` types when reading their usage. Why not make them errors?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[parser/json.go, line 170 at r7](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-MggeAKh3mxqxbNiipYp:-Mgsxzui0FuhDgaKT0yZ:b-4enmvd) ([raw file](https://github.com/m-lab/traceroute-caller/blob/c904b992180d731601840d82ae1ca84af3e26397/parser/json.go#L170)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Some of this parser code may have been copied from the etl parser and contain more general logic than required for the output created by traceroute-caller directly. Does the current implementation of traceroute-caller create JSONL output with a fixed number of objects? e.g. 4? Is 3 an error?

If this is necessary, is the UUID field guaranteed to be first?
</blockquote></details>

As discussed offline, I'd also like to simplify this.  But let me check with Greg just to be sure and send a separate PR for this.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[parser/json.go, line 180 at r7](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-Mggf9064SulPy38h3v2:-MgsyFEq5R3nNbCRc06u:bb8n3qa) ([raw file](https://github.com/m-lab/traceroute-caller/blob/c904b992180d731601840d82ae1ca84af3e26397/parser/json.go#L180)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Again, if the current tool guarantees and output format, then this indexing could be simplified hopefully
</blockquote></details>

Please see my previous comment.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
Yes, this is unfortunately a large PR because we didn't fully review and merge the individual commits.  As discussed offline, I will categorize the changes and will send smaller PRs.





<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 199 at r7](https://reviewable.io/reviews/m-lab/traceroute-caller/117#-Mgg_ftV2ZQpF_PUkhOz:-MgsxqUp4LDQIs0gawyW:b-896fix) ([raw file](https://github.com/m-lab/traceroute-caller/blob/c904b992180d731601840d82ae1ca84af3e26397/hopannotation/hopannotation.go#L199)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

When combining errors, I believe it's preferable to use `%w` so that callers (within this package) could use `errors.Is` to tell if this is a specific error derived from `errCreatePath`.

EDIT: I see that `errCreatePath` is a string... The naming made me expect that it was an error type.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

Commit:
Make final changes before splitting into smaller PRs

This commit includes changes for:

  - Add UUID line to the test JSONL files in the parser package so we
    can simplify the parser code (i.e., always expects four lines in a
    traceroute output file).
  - Use different directory paths for traceroute and hopannotation1.
  - Reset hop cache at midnight.
  - Parse cycle-start time and use it as the timestamp for hopannotation1
    datatype.

Tested locally via "go test ./..." and "docker-compose up".

Commit:
Update hopannotation1 struct field names

SaiedKazemi commented:
Closing this pull request because it has become too large.  Will re-partition the changes and send smaller pull requests.
